compiler/rustc_trait_selection/src/solve/trait_goals.rs

```diff
@@ -482,74 +491,90 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
 
         let tcx = ecx.tcx();
 
-        let a_ty = goal.predicate.self_ty();
-        let b_ty = goal.predicate.trait_ref.args.type_at(1);
-        let ty::Dynamic(a_data, a_region, ty::Dyn) = *a_ty.kind() else {
-            return vec![];
-        };
-        let ty::Dynamic(b_data, b_region, ty::Dyn) = *b_ty.kind() else {
-            return vec![];
-        };
+        // Need to wrap in a probe since `normalize_non_self_ty` has side-effects.
+        ecx.probe(|_| CandidateKind::DynUpcastingAssembly).enter(|ecx| {
```

Comment:
nit: this probe feels quite complex now :thinking: is there something in here that we can move into a separate function?

also realized that 
```rust
trait Trait {
    fn foo(&self);
}

fn foo(x: Box<dyn Trait + Send>) -> Box<dyn Send> {
    x
}
```
doesn't compile, this feels like an oversight to me? :thinking: i would except us to accept this :thinking: at least unstably